enzyme/Enzyme/CallDerivatives.cpp

```diff
@@ -1150,7 +1150,8 @@ void AdjointGenerator::handleMPI(llvm::CallInst &call, llvm::Function *called,
           C = CE->getOperand(0);
         }
         if (auto GV = dyn_cast<GlobalVariable>(C)) {
-          if (GV->getName() == "ompi_mpi_op_sum") {
+          if (GV->getName() == "ompi_mpi_op_sum" ||
+              GV->getName() == "RSMPI_SUM") {
```

Comment:
I don't know how to resolve this.
```
source_id: DefId(0:12 ~ dot_enzyme[f53a]::dot_parallel)
num_fnc_args: 5
input_activity.len(): 5
error: <unknown>:0:0: in function preprocess__ZN10dot_enzyme12dot_parallel17hd37f1f8a2c8de07dE double (ptr, ptr, i64, ptr, i64): Enzyme: cannot compute with global variable that doesn't have marked shadow global
@RSMPI_SUM = external local_unnamed_addr global ptr
```

This is the relevant code: https://github.com/rsmpi/rsmpi/commit/840e01c9f772806a7b48fd82fcacba062a2c61eb#diff-9a676b0d0c142cd1e89e8174ddb007db982d8602bd374a04e40e9f6a421acaebR216-R228

Run with
```console
$ RUSTFLAGS='-Z unstable-options' cargo +enzyme r --example=dot_enzyme --release
```

Reply:
Me neither, but can you share the module you got from ENZYME_OPT=1? Then I can experiment around to see if I find the right changes.

Reply:
@jedbrown you need to add RSMPI_SUM to the ActivityAnalysis.cpp code [the message (poorly) warns that the global variable is differentiable, but Enzyme is unable to determine a differentiable version of the global. Of course it makes no sense to differentiate wrt MPI_SUM so we can mark that in ActivityAnalysis.cpp]

Reply:
Thanks. Why didn't `ompi_mpi_op_sum` need to be there?

Reply:
I think one has an extra pointer indirection causing a load which needs to be analyzed

Reply:
We're now at this error.
```
source_id: DefId(0:12 ~ dot_enzyme[8922]::dot_parallel)
num_fnc_args: 5
input_activity.len(): 5
error: <unknown>:0:0: in function preprocess__ZN10dot_enzyme12dot_parallel17hf4631abf802872b2E double (ptr, ptr, i64, ptr, i64): Enzyme:  call:   %23 = call noundef i32 @MPI_Allreduce(ptr noundef nonnull %7, ptr noundef nonnull %6, i32 noundef 1, ptr noundef %9, ptr noundef %10, ptr noundef %22) #111, !noalias !800
 unhandled mpi_allreduce op:   %10 = load ptr, ptr @RSMPI_SUM, align 8, !noalias !803, !noundef !806
```

Reply:
yeah in this case, now the issue is that the op of MPI_allreduce cannot be detected (the earlier check if the argument was a literal global no longer applies, since this is a load of RSMPI_SUM). Changing the MPI_Allreduce check to consider something along the lines of:

```
if (LI = dyn_cast<LoadInst>(...)))
  if (auto GV = dyn_cast<GlobalVariable>(LI->getPointerOperand()))
    if (GV->getName() == "RSMPI_SUM")
      legal = true;
```

Reply:
I pushed that (assuming it was the correct place), but now I have
```
source_id: DefId(0:12 ~ dot_enzyme[42dc]::dot_parallel)
num_fnc_args: 5
input_activity.len(): 5
error: <unknown>:0:0: in function preprocess__ZN10dot_enzyme12dot_parallel17h2f3ed146b457ca09E double (ptr, ptr, i64, ptr, i64): Enzyme: cannot compute with global variable that doesn't have marked shadow global
@RSMPI_COMM_WORLD = external local_unnamed_addr global ptr
```

Reply:
okay that's the same as the first issue [a global which cannot be proven non-differentiable]. ActivityAnalysis.cpp is again the right place to add that.

Reply:
Thanks for the hand-holding. I think this output is correct now.
```console
$ RUSTFLAGS='-Z unstable-options' cargo +enzyme mpirun -n 2 --example=dot_enzyme --release | sort
[0] bx: [0.0, 2.0, 4.0, 6.0, 8.0], by: [0.0, 2.0, 4.0, 6.0, 8.0]
[0] local: 30
[1] bx: [200.0, 202.0, 204.0, 206.0, 208.0], by: [20.0, 22.0, 24.0, 26.0, 28.0]
[1] local: 6130
global: 6160
global: 6160
```

Reply:
Note that `--release` is required, otherwise I see
```
invertedPointers:
   invertedPointers[ptr %0] =   <badref> = load ptr, ptr %4, align 8
end invertedPointers
  <badref> = load ptr, ptr %4, align 8
rustc: /home/jed/src/rust-enzyme/src/tools/enzyme/enzyme/Enzyme/GradientUtils.cpp:8489: virtual void InvertedPointerVH::deleted(): Assertion `0 && "erasing something in invertedPointers map"' failed.
error: could not compile `mpi` (example "dot_enzyme")
```